Fix braindead CSVDiff

milljm opened the pull request:
It wasn't doing anything except making sure you wern't comparing apples to rocks. Modify our custom operator to actually do work.

Create Unittests, in the hopes it will never go unchecked again.

Added new feature to test absolute differences. This requires relative error/tolerance to be set at zero in order to trigger new behavior, as so:

```conf
[Test]
  type = (All SchemaDiff Types including CSVDIFF)
  rel_err = 0
  abs_err = 5.5e-6

  # Or a specific field (CSV type tests only)
  type = CSVDIFF
  override_columns = z_field  #
  override_rel_err = 0        #
  override_abs_err = 5.5e-6   # for absolute diffing on z_field
[]
```

Closes #24828

Commit:
Fix braindead CSVDiff

It wasn't doing anything except making sure you wern't comparing
apples to rocks. Modify our custom operator to actually do work.

Create Unittests, in the hopes it will never go unchecked again.

Added new feature to test absolute differences. This requires
relative error/tolerance to be set at zero in order to trigger
new behavior, as so:

```conf
[Test]
  type = (All SchemaDiff Types including CSVDIFF)
  rel_err = 0
  abs_err = 5.5e-6

  # Or a specific field (CSV type tests only)
  type = CSVDIFF
  override_columns = z_field  #
  override_rel_err = 0        #
  override_abs_err = 5.5e-6   # for absolute diffing on z_field
[]
```

Closes #24828

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 65977abfe982c9b00a93eaa52e47c169d9c28085
+Subproject commit bd07d54c44b9ee8ff381707e1117e365d8079096
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1626240/) on 75430b8 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24854/site)

This comment will be updated on new commits.

loganharbour commented:
Did you mean to update large_media?

Commit:
Revert "Revert "CSVDiff to SchemaDiff""

This reverts commit ee1fcfd051758686d0609d62179a5d54cb6313c7.

milljm commented on test/tests/test_harness/csvdiff_comparison_floor.cmp:
```diff
@@ -1,7 +1,7 @@
 
 GLOBAL VARIABLES relative 5.5e-06 floor 1e-11
-    z_field floor 1e+01        # min: 0.000e+00 @ t0          max: 4.025e-01 @ t5
+    z_field absolute 1e+01     # min: 0.000e+00 @ t0          max: 4.025e-01 @ t5
```
hmmm... this doesn't look right.

Commit:
Modify comparison files

Modify unittest comparison files so that only the thing we are
trying to test is present (make less confusing). In order to do
so, allow applying defaults for things not set.

Commit:
Clean up the schemadiff custom_operator

This hopfully fixes all the things.

Closes #23522

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1639663/) on c3977fb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>be3816</th>
      <th colspan=4>#24854 c3977f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.56%</td>
      <td>85.56%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89251</td>
      <td>89247</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15058</td>
      <td>15062</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24854/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24854/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24854/coverage/xfem)

</details>



This comment will be updated on new commits.

loganharbour commented on python/TestHarness/testers/CSVDiff.py:
```diff
@@ -7,114 +7,258 @@
+        value_id = 0
+        value = ''
+        for index, column in enumerate(orig):
+            _rel_err = abs(rel_err)
+            _abs_err = abs(abs_err)
+            _abs_zero = abs(abs_zero)
```
Why underscore prefix here? This is not intuitive to me, as we usually use this prefix for member vars. I would prefer a name that provides more context

loganharbour reviewed (changes requested):
Looks great to me. the `do_xxx` methods in the operator make life nice

Commit:
Syntax, Grammar, minor fixes

Address poorly named variables.

Separate type change for two variables into two try statements incase
one fails and we miss type changing the second variable.

Doctsring tweaks...

loganharbour commented on python/TestHarness/testers/SchemaDiff.py:
```diff
@@ -30,7 +34,7 @@ def validParams():
         params.addParam('rel_err',      5.5e-6, 'Relative error value allowed in comparisons. If '
```
Remove rel_err and abs_zero because they come from FileTester

Commit:
Adjust tolerance params

Remove the unnecessary params which are being inherited.

Match absolute error tolerance with relative error tolerance.

Fix printout (time step vs CSV row output). Also make it pretty.

Closes #24828

loganharbour commented on python/TestHarness/testers/SchemaDiff.py:
```diff
@@ -151,8 +154,29 @@ def format_diff(self, diff_dict):
+
         return formated_results
 
+    def compare_func(self, x, y, level):
+        """
+        Our custom iterator. Needed for what we believe is a bug
+        See issue: https://github.com/seperman/deepdiff/issues/404
+        """
+        return True
```
I think this is my only remaining concern. I don't want to change the behavior for JSONDiff/XMLDiff unless we really know it's a problem there too. In those cases, it's a more complex diff so maybe it doesn't suffer the same issues?

milljm replied:
I'll make the change, but consider this:

Wouldn't the same bug exhibit itself in json/xml data? I am passing the same data type that JSON does when performing CSV tests (dictionary of lists).

loganharbour replied:
Maybe. I'm just very hesitant to change everything but Exodiff in this PR
